# Telcom Customer Churn Prediction - Exploratory Data Analysis

This notebook provides a comprehensive exploratory data analysis and data preparation for the telcom customer churn prediction project.

## Project Overview
- **Dataset**: Telcom Customer Churn Dataset
- **Objective**: Analyze customer behavior patterns and prepare data for churn prediction modeling
- **Target Variable**: Churn (Yes/No)

## Notebook Structure
1. Data Loading and Initial Inspection
2. Data Quality Analysis
3. Exploratory Data Analysis - Target Variable
4. Exploratory Data Analysis - Numeric Features
5. Exploratory Data Analysis - Categorical Features
6. Data Cleaning and Preprocessing
7. Feature Engineering
8. Data Encoding
9. Final Dataset Preparation

## 1. Data Loading and Initial Inspection

Let's start by importing the necessary libraries and loading the dataset.

In [1]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

# Set random seed for reproducibility
np.random.seed(42)

# Configure plotting style
plt.style.use('default')
sns.set_palette("husl")
warnings.filterwarnings('ignore')

# Display settings
pd.set_option('display.max_columns', None)
pd.set_option('display.width', None)

In [2]:
# Load the dataset
df = pd.read_csv('Telco_Customer_Churn.csv')

print(f"Dataset shape: {df.shape}")
print(f"Columns: {list(df.columns)}")

Dataset shape: (7043, 21)
Columns: ['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents', 'tenure', 'PhoneService', 'MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn']


In [3]:
# Display first few rows
print("First 5 rows of the dataset:")
df.head()

First 5 rows of the dataset:


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

In [4]:
# Dataset information
df.info()
print("Data Types:")
print(df.dtypes)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [5]:
# Descriptive statistics
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

### **Initial Data Inspection Insights**

**Key Findings from Initial Analysis:**
- **Dataset Size**: 7,043 customers with 21 features
- **Data Quality**: All columns have complete records (no null values initially)
- **Feature Mix**: 18 categorical features (object type), 2 numeric integers, 1 float
- **Notable Issue**: `TotalCharges` is stored as object (text) instead of numeric - this needs investigation

**Customer Demographics Summary:**
- **Senior Citizens**: ~16.2% of customers are senior citizens
- **Tenure**: Average customer tenure is ~32 months (ranging from 0-72 months)
- **Monthly Charges**: Average $64.76/month (ranging from $18.25-$118.75)
- **Tenure Distribution**: Median tenure is 29 months, indicating a fairly balanced customer base

**Next Steps**: Need to investigate the TotalCharges column and check for data quality issues.

## 2. Data Quality Analysis

Let's examine the data quality by checking for duplicates, missing values, and data inconsistencies.

In [ ]:
# Check for duplicate records
print(f"Total rows: {len(df)}")
print(f"Unique rows: {len(df.drop_duplicates())}")
print(f"Duplicate rows: {len(df) - len(df.drop_duplicates())}")

# Check for duplicate customerIDs
duplicate_customers = df['customerID'].duplicated().sum()
print(f"Duplicate customer IDs: {duplicate_customers}")

In [ ]:
# Check TotalCharges for data quality issues (spaces instead of numbers)
print(f"TotalCharges data type: {df['TotalCharges'].dtype}")
print(f"Sample TotalCharges values:")
print(df['TotalCharges'].head(10))

# Check for non-numeric values in TotalCharges
non_numeric_total_charges = df[pd.to_numeric(df['TotalCharges'], errors='coerce').isna()]
print(f"\nRows with non-numeric TotalCharges: {len(non_numeric_total_charges)}")

if len(non_numeric_total_charges) > 0:
    print("Non-numeric TotalCharges values:")
    print(non_numeric_total_charges[['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges']].head())

### **Data Quality Analysis Insights**

- **No Duplicate Records**
- **No Missing Values**

**Critical Data Quality Issue Identified:**
- **TotalCharges Problem**: 11 customers have **empty space characters** instead of numeric values
- **Pattern Identified**: All problematic records have `tenure = 0` (new customers)
- **Root Cause**: New customers (0 tenure) should have $0 total charges, but data entry used spaces

**Impact Assessment:**
- **Affected Records**: 11 out of 7,043 (0.16% of dataset)
- **Business Logic**: New customers (0 tenure) logically should have $0 total charges
- **Data Type Issue**: TotalCharges stored as text instead of numeric prevents mathematical operations

**Cleaning Strategy:**
1. Convert TotalCharges to numeric format
2. Replace space characters with 0 for new customers
3. Validate that tenure=0 customers have appropriate total charges